### ISSUE TITLE
Symbols not being found during linking phase when building with the mold linker

### ISSUE THREAD

**Issue by kostadinsh** [libfuse/libfuse#810]
fuse has an issue regarding symbols not being found during the linking phase when using the mold linker. Below are the last couple of lines from the build logs, and I am also adding the complete logs as an attachment.

fuse source is at 1cb6e17e0c83b38729b5d71bfc8d7726dc3dbb61
mold used is at commit b04aba89d3a1931470983212925443e7aefca1e1
gcc (Gentoo 13.1.1_p20230527 p3) 13.1.1 20230527

Steps to reproduce:
1. clone this repo and `cd` into the libfuse folder
2. run `export LDFLAGS="-fuse-ld=mold"`
3. run `meson --prefix=/usr build`
4. run `meson compile -vC build` 

```
[48/73] cc  -o lib/libfuse3.so.3.15.0 lib/libfuse3.so.3.15.0.p/fuse.c.o lib/libfuse3.so.3.15.0.p/fuse_loop.c.o lib/libfuse3.so.3.15.0.p/fuse_loop_mt.c.o lib/libfuse3.so.3.15.0.p/fuse_lowlevel.c.o lib/libfuse3.so.3.15.0.p/fuse_opt.c.o lib/libfuse3.so.3.15.0.p/fuse_signals.c.o lib/libfuse3.so.3.15.0.p/buffer.c.o lib/libfuse3.so.3.15.0.p/cuse_lowlevel.c.o lib/libfuse3.so.3.15.0.p/helper.c.o lib/libfuse3.so.3.15.0.p/modules_subdir.c.o lib/libfuse3.so.3.15.0.p/mount_util.c.o lib/libfuse3.so.3.15.0.p/fuse_log.c.o lib/libfuse3.so.3.15.0.p/compat.c.o lib/libfuse3.so.3.15.0.p/mount.c.o lib/libfuse3.so.3.15.0.p/modules_iconv.c.o -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,--start-group -Wl,-soname,libfuse3.so.3 -fuse-ld=mold -Wl,--version-script,/home/kostadin/libfuse/lib/fuse_versionscript -pthread -ldl -lrt -Wl,--end-group
mold: warning: /home/kostadin/libfuse/lib/fuse_versionscript: cannot assign version `FUSE_3.0` to symbol `fuse_register_module`: symbol not found
mold: warning: /home/kostadin/libfuse/lib/fuse_versionscript: cannot assign version `FUSE_3.4` to symbol `fuse_reply_copy_file_range`: symbol not found
```

When using -flto even more symbols are not being found
Steps to reproduce:
1. clone this repo and `cd` into the libfuse folder
2. run `export CFLAGS="-O2 -flto"`
3. run `export LDFLAGS="-fuse-ld=mold"`
4. run `meson --prefix=/usr build`
5. run `meson compile -vC build` 

```
48/73] cc  -o lib/libfuse3.so.3.15.0 lib/libfuse3.so.3.15.0.p/fuse.c.o lib/libfuse3.so.3.15.0.p/fuse_loop.c.o lib/libfuse3.so.3.15.0.p/fuse_loop_mt.c.o lib/libfuse3.so.3.15.0.p/fuse_lowlevel.c.o lib/libfuse3.so.3.15.0.p/fuse_opt.c.o lib/libfuse3.so.3.15.0.p/fuse_signals.c.o lib/libfuse3.so.3.15.0.p/buffer.c.o lib/libfuse3.so.3.15.0.p/cuse_lowlevel.c.o lib/libfuse3.so.3.15.0.p/helper.c.o lib/libfuse3.so.3.15.0.p/modules_subdir.c.o lib/libfuse3.so.3.15.0.p/mount_util.c.o lib/libfuse3.so.3.15.0.p/fuse_log.c.o lib/libfuse3.so.3.15.0.p/compat.c.o lib/libfuse3.so.3.15.0.p/mount.c.o lib/libfuse3.so.3.15.0.p/modules_iconv.c.o -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,--start-group -Wl,-soname,libfuse3.so.3 -fuse-ld=mold -O2 -flto -Wl,--version-script,/home/kostadin/libfuse/lib/fuse_versionscript -pthread -ldl -lrt -Wl,--end-group
mold: warning: /home/kostadin/libfuse/lib/fuse_versionscript: cannot assign version `FUSE_3.0` to symbol `fuse_loop_mt`: symbol not found
mold: warning: /home/kostadin/libfuse/lib/fuse_versionscript: cannot assign version `FUSE_3.0` to symbol `fuse_session_loop_mt`: symbol not found
mold: warning: /home/kostadin/libfuse/lib/fuse_versionscript: cannot assign version `FUSE_3.0` to symbol `fuse_parse_cmdline`: symbol not found
mold: warning: /home/kostadin/libfuse/lib/fuse_versionscript: cannot assign version `FUSE_3.0` to symbol `fuse_new`: symbol not found
mold: warning: /home/kostadin/libfuse/lib/fuse_versionscript: cannot assign version `FUSE_3.0` to symbol `fuse_register_module`: symbol not found
mold: warning: /home/kostadin/libfuse/lib/fuse_versionscript: cannot assign version `FUSE_3.1` to symbol `fuse_new`: symbol not found
mold: warning: /home/kostadin/libfuse/lib/fuse_versionscript: cannot assign version `FUSE_3.2` to symbol `fuse_session_loop_mt`: symbol not found
mold: warning: /home/kostadin/libfuse/lib/fuse_versionscript: cannot assign version `FUSE_3.2` to symbol `fuse_loop_mt`: symbol not found
mold: warning: /home/kostadin/libfuse/lib/fuse_versionscript: cannot assign version `FUSE_3.4` to symbol `fuse_reply_copy_file_range`: symbol not found
mold: warning: /home/kostadin/libfuse/lib/fuse_versionscript: cannot assign version `FUSE_3.12` to symbol `fuse_session_loop_mt`: symbol not found
mold: warning: /home/kostadin/libfuse/lib/fuse_versionscript: cannot assign version `FUSE_3.12` to symbol `fuse_loop_mt`: symbol not found
mold: warning: /home/kostadin/libfuse/lib/fuse_versionscript: cannot assign version `FUSE_3.12` to symbol `fuse_parse_cmdline`: symbol not found
lto-wrapper: warning: using serial compilation of 2 LTRANS jobs
lto-wrapper: note: see the ‘-flto’ option documentation for more information
mold: warning: /home/kostadin/libfuse/lib/fuse_versionscript: cannot assign version `FUSE_3.0` to symbol `fuse_register_module`: symbol not found
mold: warning: /home/kostadin/libfuse/lib/fuse_versionscript: cannot assign version `FUSE_3.4` to symbol `fuse_reply_copy_file_range`: symbol not found
```

[fuse-nolto.log](https://github.com/libfuse/libfuse/files/11936371/fuse-nolto.log)
[fuse-lto.log](https://github.com/libfuse/libfuse/files/11936372/fuse-lto.log)

**Comment by Nikratio** [comment]
Thanks for taking the time to report this issue!

Unfortunately, this project does not currently have any active, regular contributors. As the maintainer, I try to review pull requests and make regular releases, but unfortunately I have no capacity to do significant development beyond that.
Issue reports that do not come with a pull request or clearly have high impact on a large number of users are therefore likely to languish.

I understand that this is frustrating for users, but I hope you can also understand that any development work that I do on this project has to compete with spending time with my family, doing work that I get paid for, doing something recreational without a computer, or working on features/bugs that affect me personally. Most bugs and ideas - unfortunately including this one - loose out in this competition.

In other words, unless you plan to work on this yourself or can recruit someone who will, it's unlikely that anyone is going to do anything about it anytime soon.

This is just to calibrate expectations. I am grateful that you took the time and effort to report this! I'll leave this issue open to document the problem and who knows, maybe someone will pick it up after all :-).

**Comment by bsbernd** [comment]
Looks like a symver issue with mold -  @kostadinsh check with mold developers please. Also, symver can be disabled in libfuse meson options.

**Comment by bsbernd** [comment]
fuse_register_module() is a static function - either static needs to be removed or it not be in lib/fuse_versionscript. Changed in commit d4e294b79. Given this commit predates fuse-3.0, I think we should be good to remove the symbol from the version script.


fuse_reply_copy_file_range() is not defined/declared at all - bug in commit fe4f9428fc403fa8b99051f52d84ea5bd13f3855. I can't decide on the first glance if the function is missing or should be removed from the version script. High level API uses fuse_lib_copy_file_range (in fuse_lib_copy_file_range()), I'm going to check kernel code later on to see if that if sufficient.

So that for the report, the remaining mold errors are correct.

**Comment by bsbernd** [comment]
@kostadinsh Could you test the patch above? I use mold myself for some projects, but I don't have the time today to test it with libfuse.

**Comment by kostadinsh** [comment]
> @kostadinsh Could you test the patch above? I use mold myself for some projects, but I don't have the time today to test it with libfuse.

Just tested it, and it builds properly. Thanks!
